avro-schema-processor/src/main/java/io/micronaut/avro/visitor/AvroSchemaVisitor.java

```diff
@@ -0,0 +1,398 @@
+            if (isPrimitiveType(property.getType()) || isPrimitiveAvroType(propertySchema.getType()) && !isLogicalAvroType(propertySchema.getLogicalType()) && !propertySchema.isUnsupported() || propertySchema.isRefType()) {
+                field.setType(propertySchema.getType());
+            } else {
+                field.setType(propertySchema);
+            }
+
+            avroSchema.addField(field);
+        }
+    }
+
+    private static void writeSchema(AvroSchema avroSchema, ClassElement originatingElement, VisitorContext visitorContext, AvroSchemaContext context) {
+        String fileName = getFileName(avroSchema);
+        String path = context.outputLocation() + SLASH + fileName;
+        GeneratedFile specFile = visitorContext.visitMetaInfFile(path, originatingElement).orElse(null);
+        if (specFile == null) {
+            visitorContext.warn("Unable to get [\" " + path + "\"] file to write Avro schema", null);
+        } else {
+            visitorContext.info("Generating Avro schema file: " + specFile.getName());
+            try (Writer writer = specFile.openWriter()) {
+                ObjectMapper mapper = ObjectMapper.getDefault();;
+                List<AvroSchema.Field> sortedFields = avroSchema.getFields().stream()
+                    .sorted(Comparator.comparing(AvroSchema.Field::getName))
+                    .collect(Collectors.toList());
+                avroSchema.setFields(sortedFields);
+                String schema = mapper.writeValueAsString(avroSchema);
+                writer.write(schema);
+            } catch (IOException e) {
+                throw new RuntimeException("Failed writing Avro schema " + specFile.getName() + " file: " + e, e);
+            }
+        }
+    }
+
+    private static String getFileName(AvroSchema avroSchema) {
+        String fileName;
+        if (avroSchema.getFullName() != null) {
+            fileName = avroSchema.getFullName();
+        } else {
+            fileName = avroSchema.getName();
+        }
+        fileName = fileName.replace('.', '/');
+        return fileName + "-schema" + SUFFIX;
+    }
+
+    private static boolean isLogicalAvroType(LogicalType type) {
+        return (type == LogicalType.DATE || type == LogicalType.UUID || type == LogicalType.DURATION ||
+            type == LogicalType.DECIMAL || type == LogicalType.TIME_MILLIS || type == LogicalType.TIME_MICROS ||
+            type == LogicalType.TIMESTAMP_MILLIS || type == LogicalType.TIMESTAMP_MICROS);
+    }
+
+    private static Type getPrimitiveAvroType(String typeName) {
+        return switch (typeName) {
+            case "int" -> Type.INT;
+            case "float" -> Type.FLOAT;
+            case "long" -> Type.LONG;
+            case "double" -> Type.DOUBLE;
+            case "boolean" -> Type.BOOLEAN;
+            case "void" -> Type.NULL;
+            default ->
+                throw new IllegalArgumentException("Unsupported primitive type: " + typeName);
```

Comment:
this should be a `ProcessingException` and you should pass the class element / method element / field element etc being visited